modules/mod-deep-learning/ModelCard.cpp

```diff
@@ -0,0 +1,371 @@
+{
+   rapidjson::StringBuffer sb;
+   rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+
+   writer.StartObject();
+   Serialize(writer);
+   writer.EndObject();
+
+   wxFile file(path, wxFile::write);
+   if (!file.Write(wxString(sb.GetString())))
+      throw InvalidModelCardDocument(XO("Could not serialize ModelCard to file"), "");
+}
+
+void ModelCard::DeserializeFromFile(const std::string &path, const Doc& schema)
+{
+   Doc d = parsers::ParseFile(path);
+   Deserialize(d, schema);
+}
+
+template < typename Writer >
+void ModelCard::Serialize(Writer &writer) const
+{
+   // name
+   writer.String("name");
+   writer.String(m_name.c_str());
+
+   // author
+   writer.String("author");
+   writer.String(m_author.c_str());
+   
+   // long description
+   writer.String("long_description");
+   writer.String(m_long_description.c_str());
+
+   // short description
+   writer.String("short_description");
+   writer.String(m_short_description.c_str());
+
+   // sample rate
+   writer.String("sample_rate");
+   writer.Int(m_sample_rate);
+
+   // multichannel
+   writer.String("multichannel");
+   writer.Bool(m_multichannel);
+
+   // effect type
+   writer.String("effect_type");
+   writer.String(m_effect_type.c_str());
+
+   // domain tags
+   writer.String("domain_tags");
+   writer.StartArray();
+   for (auto tag : m_domain_tags)
+      writer.String(tag.c_str());
+   writer.EndArray();
+
+   // tags
+   writer.String("tags");
+   writer.StartArray();
+   for (auto tag : m_tags)
+      writer.String(tag.c_str());
+   writer.EndArray();
+
+   // labels
+   writer.String("labels");
+   writer.StartArray();
+   for(auto label : m_labels)
+      writer.String(label.c_str());
+   writer.EndArray();
+
+   // model size
+   writer.String("model_size");
+   writer.Uint64((uint64_t)m_model_size);
+
+}
+
+void ModelCard::Deserialize(const Doc& doc, const Doc& schema)
+{
+   using namespace validators;
+   try
+   {
+      Validate(doc, schema);
+   }
+   catch(const InvalidModelCardDocument& e)
+   {
+      wxLogError(wxString(e.what()));
+      return;
+   }
+
+   // these fields are not in HF mettadata but rather added later,
+   // so don't throw if they are not present
+   m_author = tryGetString("author", doc, "");
+   m_name = tryGetString("name", doc, "");
+   m_model_size = (size_t)tryGetUint64("model_size", doc, 0);
+
+   m_long_description = tryGetString("long_description", doc, 
+                                       "no long description available");
+   m_short_description = tryGetString("short_description", doc, 
+                                       "no short description available");
+   m_effect_type = toLowerCase(tryGetString("effect_type", doc));
+   m_domain_tags = tryGetStringArray("domain_tags", doc, true);
+   m_tags = tryGetStringArray("tags", doc);
+   m_labels = tryGetStringArray("labels", doc);
+   m_sample_rate = tryGetInt("sample_rate", doc);
+   m_multichannel = tryGetBool("multichannel", doc);
```

Comment:
Permuting these lines so the order agrees with the serialization function would help a reviewer verify it's complete.

Reply:
resolved in 1a5105597